### ISSUE TITLE
More details on migration to Drupal 8

### ISSUE THREAD

**Issue by stevector** [pantheon-systems/documentation#1386]
Re: https://pantheon.io/docs/articles/drupal/major-version-drupal-upgrades/

 ## Issue Description

We could benefit from more detail regarding upgrading/migrating to Drupal 8, especially for a site that is already on Pantheon. Particularly we could use information on
- DNS considerations for a site that is live on Pantheon using Drupal 6 or 7.
- Connecting to a source database with migrate module.
- Migrating uploaded files.
## Suggested Resolution

Do a test migration of a Drupal 6 or 7 site to Drupal 8 and write down the steps in greater detail.

**Comment by stevector** [comment]
I'm working on a migration of a personal project from D7 to D8. To set up the migration connection info I am using the terminus secrets plugin that @greg-1-anderson wrote: https://github.com/pantheon-systems/terminus-secrets-plugin With this plugin I can put the DB connection from the D7 site in a secrets file on the D8 site.

```
terminus secrets set --env=$CIRCLE_BUILD_NUM migrate_source_db__database $(terminus site  connection-info  --site=nerdologues  --env=migrateprep  --field=mysql_database)
terminus secrets set --env=$CIRCLE_BUILD_NUM migrate_source_db__username $(terminus site  connection-info  --site=nerdologues  --env=migrateprep  --field=mysql_username)
terminus secrets set --env=$CIRCLE_BUILD_NUM migrate_source_db__password $(terminus site  connection-info  --site=nerdologues  --env=migrateprep  --field=mysql_password)
terminus secrets set --env=$CIRCLE_BUILD_NUM migrate_source_db__host     $(terminus site  connection-info  --site=nerdologues  --env=migrateprep  --field=mysql_host)
terminus secrets set --env=$CIRCLE_BUILD_NUM migrate_source_db__port     $(terminus site  connection-info  --site=nerdologues  --env=migrateprep  --field=mysql_port)
```
In that example I'm making a new multidev environment per Circle CI build, which is probably overkill for many use cases. The relevant piece is that DB credentials on the source DB (any DB on Pantheon) can change at any time. So in situations where a migration is going to run continually over a long period of time, it would be a good idea to somehow refresh the source DB info.


Then in a file `settings.migrate-on-pantheon.php` I have 

```php

/**
 * Get secrets from secrets file.
 *
 * @param array $requiredKeys  List of keys in secrets file that must exist.
 */
// These secret keys come from the script deploy-to-multidev. 
$required_keys = [
  'migrate_source_db__database',
  'migrate_source_db__username',
  'migrate_source_db__password',
  'migrate_source_db__host',
  'migrate_source_db__port'
];
$defaults= array();
$get_secrets = function ($required_keys, $defaults)
{
  $secretsFile = $_SERVER['HOME'] . '/files/private/secrets.json';
  if (!file_exists($secretsFile)) {
    die('No secrets file found. Aborting!');
  }
  $secretsContents = file_get_contents($secretsFile);
  $secrets = json_decode($secretsContents, 1);
  if ($secrets == FALSE) {
    die('Could not parse json in secrets file. Aborting!');
  }
  $secrets += $defaults;
  $missing = array_diff($required_keys, array_keys($secrets));
  if (!empty($missing)) {
    die('Missing required keys in json secrets file: ' . implode(',', $missing) . '. Aborting!');
  }
  return $secrets;
};
$secrets = $get_secrets($required_keys, $defaults);
if (!empty($secrets['migrate_source_db__password'])) {
  $databases['drupal_7']['default'] = array (
    'database' => $secrets['migrate_source_db__database'],
    'username' => $secrets['migrate_source_db__username'],
    'password' => $secrets['migrate_source_db__password'],
    'host' => $secrets['migrate_source_db__host'],
    'port' => $secrets['migrate_source_db__port'],
    'driver' => 'mysql',
    'prefix' => '',
    'collation' => 'utf8mb4_general_ci',
  );
}

```

@joshkoenig, @populist, @bmackinney what are your thoughts on this technique for connecting across sites? I was talking with @ryanweal at DrupalCon about getting Pantheon-to-Pantheon migrations documented in advance of an eventual migration-focused webinar.

**Comment by joshkoenig** [comment]
I recommend having one secret: a `machine_token` and fetching the DB connection info in real-time. That will prevent things from breaking if containers shift around.

**Comment by ryanweal** [comment]
This is great, thanks. I will do a test run sometime this week.

**Comment by greg-1-anderson** [comment]
I haven't worked with migrate on Drupal 8 yet, but I would hope that it would have a hook or something that could be used to fetch the DB connection info.  Running a Terminus command from within settings.php does not seem like the most performant solution. We should experiment and document the best technique for doing this.

**Comment by stevector** [comment]
Yeah, unfortunately there is not a hook per se for getting DB connection info.
In the Drupal migrate IRC channel, I asked:


> I'm working on docs for migrating to Drupal 8 on Pantheon from a Drupal 6/7 site that is also on Pantheon. I have a migration working where source database credentials are set in settings.php (Or entered through the UI).
> I would like to simplify the process of setting up the database credentials such that a developer could supply simply the name of the Pantheon source site (along with an authentication step).
> Right now I'm just looking at which steps in the database connection process are flexible.
> It looks like getDatabase() and setUpDatabase are the key methods: http://cgit.drupalcode.org/drupal/tree/core/modules/migrate/src/Plugin/migrate/source/SqlBase.php#n75
> I could imagine a scenario in which the module I write would want to override those methods. But that seems difficult/impossible given how many subclasses extend from SqlBase
> It seems like my options are work in core's configuration/state system such that these methods read connection info as normal.
> Or write a patch that replaces the internals of getDatabase() with a service.
> Anyone have thoughts on what I should pursue?


Mike Ryan said: 

> It seems simplest to use state as getDatabase() is expecting
> database_state_key would be my preference, except https://www.drupal.org/node/2681869 still needs to be addressed, so migrate.fallback_state_key would be the way to go

He also recommended looking at Drush commands as a reference: 

> I assume you’re running the migrations through drush? You can see how drush migrate-upgrade sets and then removes the fallback_state_key

http://cgit.drupalcode.org/migrate_upgrade/tree/src/MigrateUpgradeDrushRunner.php#n181
http://cgit.drupalcode.org/migrate_upgrade/tree/migrate_upgrade.drush.inc#n57

So, I'll look into mechanisms for mucking with the state system in ways that don't slow down every request to the D8 site.

**Comment by greg-1-anderson** [comment]
Maybe you could call terminus and update the state via a Drush pre-migrate-upgrade hook. You'd want to have one hook for every Drush command used to kick off or get status on the migration. Hopefully, though, this hook would not call terminus for every item of data migrated, as this would slow down the migration considerably.

How will long-running migrations be managed?  Via a Drush command called from cron? If there's a migrate batch process that is run over a long period of time iteratively w/out calling through Drush, then you'd risk failure if the Pantheon db was relocated mid-migration.

**Comment by greg-1-anderson** [comment]
Here's an example Drush hook:

https://github.com/drush-ops/drush/blob/master/examples/policy.drush.inc#L83

**Comment by ryanweal** [comment]
@stevector

I had been thinking of presetting the values just using formAPI as this preserves the existing workflow documentation when using the UI. I just ran some local tests and this seems to work for me... the password field will be blank but even after a fresh install you can click right past the form:

`/**
 * Implements hook_form_alter().
 */
function migrate_presets_form_alter(&$form, \Drupal\Core\Form\FormStateInterface $form_state, $form_id) {
  if ($form_id == 'migrate_drupal_ui_form') {
    //kint($form);
    //kint($form_state);
    $form['database']['settings']['mysql']['host']['#default_value'] = 'db';
    $form['database']['settings']['mysql']['database']['#default_value'] = 'cooks_legacy';
    $form['database']['settings']['mysql']['username']['#default_value'] = 'root';
    $form['database']['settings']['mysql']['password']['#value'] = 'my-secret-pw';
    $form['database']['settings']['mysql']['advanced_options']['port']['#default_value'] = '3306';

    $form['source']['source_base_path']['#default_value'] = '/path/to/files';
  }
}
`

The form appears at /upgrade when the migrate_drupal_ui module is enabled.

**Comment by stevector** [comment]
I see there being a couple of overlapping difficulties and I think we're each addressing them partially.

1. Accounting for pantheon database connection info for the source site changing at any moment. This reality makes using the out-of-the-box browser UI hard to rely on because info entered there will become out of date eventually.
2. Security concerns about storing source site database passwords. The browser UI will put DB connection info in a config entity where it might be exported and committed to git. Not ideal.
3. If the DB source connection info is not stored in the Drupal 8 DB/YML then it is probably coded into settings.php. That's what I'm doing in a side project's D7 -> D8 migration: https://github.com/stevector/nerdologues-d8/blob/0f2dbb1158ae6ab0691f5896cca04a5565d4beb1/sites/default/settings.migrate-on-pantheon.php There I'm using the `secrets.json` file that has been a pattern in our Quicksilver examples repo: https://github.com/pantheon-systems/quicksilver-examples This approach still does not guarantee up-to-date connection info.

Greg and Josh, you're talking about running terminus from inside the Drupal 8 site, yes? So you mean adding terminus to the composer.json of the Drupal 8 site so that it could be invoked from drush commands?

As a side note, I think it is ok to have different recommendations for drush vs. browser based migrations.

**Comment by greg-1-anderson** [comment]
Yes; if you `composer require` Terminus in your site, you should be able to call it from Pantheon using a machine token. I don't think I've done it before, but should work.

**Comment by ryanweal** [comment]
Some additional findings that should help this discussion continue... I have listed out the possible cases where the DB settings are used, and how they are used. For now that exists as a table shown here: http://dev-migratebook.pantheonsite.io/node/16

Use of migration_group to define a second or third DB/connection is documented a bit here: https://www.drupaleasy.com/blogs/ultimike/2016/04/drupal-6-drupal-81x-custom-content-migration

Lastly... some comments from Mike Ryan's post which talks about the API changes from 8.0.x to 8.1.x describe _when_ the migration_group database keys get used. This may provide some insight if you are thinking to piggyback on the instantiation to do the lookup at that time (I'm not sure how possible this would be, but it seems to correspond with what @stevector pasted in from Mike Ryan earlier):

"there is now no direct core support for persisting configuration changes for the plugins - any configuration (such as a database connection) injected into a migration plugin is only applied at runtime, and without a standard method of persisting that configuration a general-purpose tool (such as the drush commands in migrate_tools) has no way of knowing what to inject or how."
http://mikeryan.name/blog/mikeryan/migration-update-for-drupal-8-1

...the nice thing about this is that it does not run per-row... it runs when the migration is initialized. So in terms of slowdown, it should have minimal impact if you go that route.

With all that said, I think it might be cool to do something like the following... although I have not tested this and I'm not sure what I'm proposing makes sense... but it would be simple for developers if it is possible:

 - recommend users creating custom migrations make a migrate_plus.migration_group.mygroup.yml in config/install of their custom module. 
 - provide a custom DB key in the settings there:
```
shared_configuration:
  source:
    key: legacy
```
 - rather than define "legacy" in settings.php, replace the word "legacy" with the Pantheon machine/site hash/token/whatever, and then use custom code to reconcile what that means and deal with it automatically for the developer.

**Comment by ari-gold** [comment]
Another approach to the problem of dynamic DB connection information: https://www.eric.pe/terson/engineers/pantheon-switchboard

**Comment by ryanweal** [comment]
Aimee from Hook 42 had been using Pingdom to hit a Pantheon dev site to keep it awake (and presumably so the DB did not move?) during a large migration that ran well over 10h. I had sent around a PDF of some sketches she drew up for me at DrupalCon Amsterdam explaining how she did it. I can send it to others if they are interested.

**Comment by stevector** [comment]
Thanks @ryanweal for all the extra info! After reading through those links I agree that your suggestion is feasible. We can do custom handling of of a special key. To experiment with different ways of doing the migration I've set up a repo that could eventually hold a `migrate_pantheon` module. https://github.com/stevector/migrate_pantheon Right now it just holds a CircleCI testing suite. It automates the migration of D7 to a fresh D8 multidev and then tests that a node successful migrated.

As it stands now, the handling of the connection info in the Circle test has all the problems discussed above. Before, I dive into writing code to work around those problems I'd like to verify with @joshkoenig and @greg-1-anderson one more time that we should have a Drupal 8 site loading Terminus itself via Composer. I'm asking again because that seems like a very heavy solution that we haven't recommend for anything else that I'm aware of. I think it's cool, but I'm concerned it is overkill.

I think an alternative could be a Terminus plugin (or a PR to Terminus itself) that allows for source DB credentials to be passed into the D8 site every time  a migration-related drush command is called through Terminus.

The way I see it, our two basic option are:
1. Put Terminus inside of Drupal 8 so that Drupal 8 can reach out to a Drupal 7 site and grab credentials.
2. Put more logic inside of the developer's instance of Terminus so that it can give Drupal 8 the connection info for Drupal 7 with every command.

The link Ari provided does get around the problem of connection info changing. But it requires setting up another 3rd-party service and we'd still have the problem of storing the connection to the switchboard securely.

**Comment by greg-1-anderson** [comment]
If we can be reasonably confident that the db settings will be more-or-less stable during the migration --  there are no guarantees, as a db endpoint could be drained at any time -- then I think that perhaps the best thing to do would be to 'include' a file from settings.php, and put the db settings for the site to be migrated there. Then, a simple script could call terminus and write the updated db info to this file.

If the db info did change mid-migraiton, then you'd get an error message, and the migration would bail out. But restarting it would be easy, right?  Given that, it might be best to go with a lightweight / fast implementation, rather than slow down the migration substantially just to guarentee that the db info was exactly correct for every request.

These hypotheses should be born out by testing, though.

**Comment by stevector** [comment]
@greg-1-anderson So something like my above comment? https://github.com/pantheon-systems/documentation/issues/1386#issuecomment-219603596

Yes, if the db info changed mid-migration I expect the migration would just die. It could be restarted once the connection info was updated by rerunning the setup script. Someone could rerun whatever terminus script sets up the settings.php include. How about instead of rely on the the terminus secrets plugin, this hypothetical script writes `settings.migrate-on-pantheon.php` directly to `$_SERVER['HOME'] . '/files/private/` with the migration connection info as plain text?

**Comment by greg-1-anderson** [comment]
Oh yeah, I forgot where this all started. :)  I think that 'secrets' is fine, and writing straight to a private file is fine.

**Comment by ryanweal** [comment]
It dawned on me last night that there are a couple extra things that you may want to consider in a migrate_pantheon module, one of which is not in core yet:

 - People may want to merge sites together. My team has encountered sites where the main site and the blog get merged into one... so it would be nice if the pantheon_migration module supported multiple sources. Per my comments at http://dev-migratebook.pantheonsite.io/node/16 if the user supplies a custom group, that could use a different (legacy2, legacy3, legacyN) database.

 - More important: currently D7->D8 migrations are getting the passwords rehashed, which is a bug, and at NOLA we had to find a solution to this... the result is a new array of old D7 site hashes that will go into settings.php. Old hashes will be tried until a "known good" one is found and then D8 will log the user in and create a new hashed password after a successful login has occurred. Details are here: https://www.drupal.org/node/2598038 (currently RTBC).

**Comment by stevector** [comment]
I've made a GitHub repo that automates a Drush-based migration with CircleCI: https://github.com/stevector/migrate_pantheon/

I'm sure it'd be possible to show dealing with multiple sources and the password hashing problem there.

**Comment by ryanweal** [comment]
Sorry to revive an old thread, but is this method no longer possible? I tried manually entring DB credentials into a settings.php file and all I get now is "No address associated with hostname" (and yes I verified the credentials are still current in the UI). I'm not sure if I had used these credentials in settings.php before or if I only used in the UI (which was great for doing trainings at Badcamp 2016 btw!). If it is no longer possible to connect across instances, should we update the blog post I did for Pantheon?
